[INTEL-HPU] Deepseek R1 model enabling for Intel Gaudi

xuechendi opened the pull request:
updated in 03/24/2025

convert this PR as a draft and waiting for 1.21 HPU synapse release

----

Accuracy is verified as below
```
VLLM_SKIP_WARMUP=true \
PT_HPU_ENABLE_LAZY_COLLECTIVES=true \
PT_HPU_WEIGHT_SHARING=0 \
VLLM_TEST_ENABLE_EP=1 \
VLLM_MLA_DISABLE_REQUANTIZATION=1 \
bash .buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh -m deepseek-ai/DeepSeek-R1 -b 128 -l 256 -f 5 -t 8
```
![image](https://github.com/user-attachments/assets/871af252-2cbd-46ff-b8b9-d568bda9bc32)

3 options for running fp8 on HPU:
1. dynamic quant + static MOE => **default option** . Supported by public released hpu 1.20 synapse container
2. dynamic quant + dynamic MOE => enabled by "VLLM_DMOE_DYNAMIC_SCALE=1". Working with internal hpu synapse build and will release container later
3. static quant + dynamic MOE => enabled by using a converted model from original block-based quant to channel-based quant. Please refer https://github.com/HabanaAI/vllm-fork/blob/deepseek_r1/scripts/DEEPSEEK_R1_ON_GAUDI.md#option-3-run-with-static-quantization for instruction of how to convert model

xuechendi committed:
hpu deepseek enabling

Signed-off-by: Chendi Xue <chendi.xue@intel.com>
Co-authored-by: Yang Wang <yang3.wang@intel.com>
Co-authored-by: Lin Wei <wei2.lin@intel.com>
Co-authored-by: Mengni Wang <mengni.wang@intel.com>
Co-authored-by: Youlei Yang <youlei.yang@intel.com>
Co-authored-by: Kunshang Ji <kunshang.ji@intel.com>
Signed-off-by: Chendi Xue <chendi.xue@intel.com>

yangw1234 commented on vllm/attention/backends/hpu_attn.py:
```diff
@@ -8,20 +8,90 @@
+    PatchedVLLMKVCache)
+from neural_compressor.torch.algorithms.fp8_quant._quant_common.quant_config import (  # noqa: E501
+    Fp8cfg)
+from neural_compressor.torch.algorithms.fp8_quant.model_configs import (
```
should we add this to dependencies?

xuechendi replied:
> should we add this to dependencies?

Let me check with INC team.

xuechendi committed:
Enable MTP for HPU

os.environ['VLLM_CONTIGUOUS_PA'] = 'false'
os.environ['VLLM_MLA_DISABLE_REQUANTIZATION']='1'
os.environ['PT_HPU_ENABLE_LAZY_COLLECTIVES']='true'
os.environ['PT_HPU_WEIGHT_SHARING']='0'

llm = LLM(model=model_path,
            trust_remote_code=True,
            #enforce_eager=True,
            dtype="bfloat16",
            use_v2_block_manager=True,
            tensor_parallel_size=8,
            #speculative_draft_tensor_parallel_size=8,
            max_model_len=16384,
            num_scheduler_steps=1,
            distributed_executor_backend='mp',
            gpu_memory_utilization=0.8,
            kv_cache_dtype="fp8_inc",
            enable_expert_parallel=True,
            num_speculative_tokens=1,
            seed=2024)

output:
Prompt: 'Hello, my name is', Generated text: '\n\n# Dr. Kaitlin McConkey\n\n## Professional information\n\n-'
Prompt: 'The president of the United States is', Generated text: ' the head of state and head of government of the United States, indirectly elected to'
Prompt: 'The capital of France is', Generated text: ' Paris. The capital of Spain is Madrid. The capital of Italy is Rome.'
Prompt: 'The future of AI is', Generated text: " here, and it's called GPT-4. This groundbreaking technology is set to"

Signed-off-by: Chendi Xue <chendi.xue@intel.com>
Signed-off-by: Yang Wang <yang3.wang@intel.com>
Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi committed:
apply delayed_sampling

Signed-off-by: Chendi Xue <chendi.xue@intel.com>
Signed-off-by: Michal Adamczyk <madamczyk@habana.ai>
Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi committed:
Fix mypy issues

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi commented:
@mgoin @robertgshaw2-redhat @WoosukKwon @simon-mo @youkaichao @comaniac 
Would like to get some reviews, thanks

xinyu-intel committed:
Correct kvcache shape

Signed-off-by: Xinyu Chen <xichen@habana.ai>

Signed-off-by: Chendi.Xue <chendi.xue@intel.com>

xuechendi committed:
Fix unquantized FusedMOE

Signed-off-by: Chendi.Xue <chendi.xue@intel.com>

xuechendi committed:
Fix VLLM_MLA_DISABLE_REQUANTIZATION=0

Signed-off-by: Chendi.Xue <chendi.xue@intel.com>
